src/framing/plaintext.rs

```diff
@@ -145,47 +142,105 @@ impl MLSPlaintext {
+        Ok(())
     }
 
-    /// Verify the signature of the `MLSPlaintext`.
+    /// Verify the signature of an `MLSPlaintext` sent from an external party.
+    /// Returns `Ok(())` if successful or `VerificationError` otherwise.
     pub fn verify_signature(
         &self,
-        serialized_context_option: Option<Vec<u8>>,
+        serialized_context: &[u8],
         credential: &Credential,
-    ) -> bool {
-        let signature_input = MLSPlaintextTBS::new_from(&self, serialized_context_option);
-        signature_input.verify(credential, &self.signature)
+    ) -> Result<(), VerificationError> {
+        let tbs_payload = MLSPlaintextTBS::new_from(&self, Some(serialized_context));
+        tbs_payload.verify(credential, &self.signature)
     }
 
-    /// Verify the membership tag of the `MLSPlaintext`. Returns `true` if the
-    /// verification is successful and `false` otherwise.
+    /// Verify the signature of an `MLSPlaintext` sent from an external party.
+    /// Returns `Ok(())` if successful or `VerificationError` otherwise.
     pub fn verify_membership_tag(
         &self,
         ciphersuite: &Ciphersuite,
-        serialized_context: Vec<u8>,
+        serialized_context: &[u8],
         membership_key: &MembershipKey,
-    ) -> bool {
-        let tbs_payload =
-            MLSPlaintextTBSPayload::new_from_mls_plaintext(&self, Some(serialized_context));
-        let tbm_payload = MLSPlaintextTBMPayload::new(tbs_payload, &self);
-        let expected_membership_tag = &MembershipTag::new(ciphersuite, membership_key, tbm_payload);
+    ) -> Result<(), VerificationError> {
+        let tbs_payload = MLSPlaintextTBS::new_from(&self, Some(serialized_context));
+        let tbm_payload =
+            MLSPlaintextTBMPayload::new(&tbs_payload, &self.signature, &self.confirmation_tag)
+                .map_err(VerificationError::CodecError)?;
+        let expected_membership_tag =
+            &MembershipTag::new(ciphersuite, membership_key, tbm_payload)?;
+
+        // Verify the membership tag
+        if let Some(membership_tag) = &self.membership_tag {
+            if membership_tag != expected_membership_tag {
+                Err(VerificationError::InvalidMembershipTag)
+            } else {
+                Ok(())
+            }
+        } else {
+            Err(VerificationError::MissingMembershipTag)
+        }
+    }
 
+    /// Verify the signature and the membership tag of an `MLSPlaintext` sent
+    /// from a group member. Returns `Ok(())` if successful or
+    /// `VerificationError` otherwise.
+    pub fn verify_from_member(
+        &self,
+        ciphersuite: &Ciphersuite,
+        serialized_context: &[u8],
+        credential: &Credential,
+        membership_key: &MembershipKey,
+    ) -> Result<(), VerificationError> {
+        // Verify the signature first
+        let tbs_payload = MLSPlaintextTBS::new_from(&self, Some(serialized_context));
+        tbs_payload.verify(credential, &self.signature)?;
```

Comment:
So this is something that I can see leading to a timing oracle. because we return as soon as a verification fails. (This is another reason why there should be one function.) I'm fine with leaving it like this but again please add the potential issue to #133 so we don't forget about it.